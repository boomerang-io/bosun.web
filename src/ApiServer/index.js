import { Server, Serializer, Model } from "miragejs";
import { inflections } from "inflected";
import uuid from "uuid/v4";
import { serviceUrl } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("policy", "policy");
    inflect.irregular("violation", "violation");
    inflect.irregular("template", "template");
    inflect.irregular("insights", "insights");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      templates: Model,
      policies: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });
      this.get(serviceUrl.getCicdNavigation({}), (schema, request) => {
        return schema.db.cicdNavigation.map((item) => {
          if (typeof item.link === "string") {
            item.link = item.link.replace(":teamId", request.queryParams.teamId);
          }

          return item;
        });
      });

      this.get(serviceUrl.getPlatformNavigation(), (schema) => {
        return schema.db.navigation[0];
      });

      this.get(serviceUrl.getTeams(), (schema) => {
        return schema.db.teams;
      });

      this.get(serviceUrl.getPolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        return activePolicy;
      });

      this.get(serviceUrl.getValidateInfo({ policyId: ":policyId" }), (schema) => {
        return schema.db.validateInfo;
      });

      this.get(serviceUrl.getTemplates(), (schema) => {
        return schema.db.templates;
      });

      this.get(serviceUrl.getPolicyOverview(), (schema, request) => {
        let { teamId } = request.queryParams;
        return schema.db.policies.filter((policy) => policy.teamId === teamId);
      });

      this.get(serviceUrl.getViolationsOverview(), (schema) => {
        return schema.db.violations;
      });

      this.get(serviceUrl.getInsightsOverview(), (schema) => {
        return schema.db.insights;
      });

      this.post(serviceUrl.postCreatePolicy(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.policies.create({ ...body, id: uuid() });
      });

      this.post(serviceUrl.postCreatePolicyTemplate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.templates.create({ ...body, id: uuid() });
      });

      this.patch(serviceUrl.patchUpdatePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activePolicy = schema.policies.find(policyId);
        activePolicy.update(body);
        return schema.policies.all();
      });

      this.patch(serviceUrl.patchUpdatePolicyTemplate({ templateId: ":templateId" }), (schema, request) => {
        let { templateId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTemplate = schema.templates.find(templateId);
        activeTemplate.update(body);
        return schema.templates.all();
      });

      this.delete(serviceUrl.deletePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        activePolicy.destroy({ id: policyId });
        return schema.policies.all();
      });
    },
  });
}
