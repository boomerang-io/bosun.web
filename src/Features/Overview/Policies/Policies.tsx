import React from "react";
import NoDisplay from "Components/NoDisplay";
import { Button } from "@boomerang-io/carbon-addons-boomerang-react";
import { useHistory } from "react-router-dom";
import PoliciesTable from "./PoliciesTable";
import { appLink } from "Config/appConfig";
import { Add16 } from "@carbon/icons-react";
import { PolicyData } from "Types";
import styles from "./policies.module.scss";

type Props = {
  policies: PolicyData[];
  activeTeamId?: string;
};

export function Policies({ policies, activeTeamId }: Props) {
  let history = useHistory();
  return (
    <section className={styles.container} data-testid="policies-container" id="policies-container">
      <h2 className={styles.title}>{`Policies (${policies.length})`}</h2>
      <p className={styles.message}>
        All of your teams policies can be found here. Go forth and define rules to enforce your standards!
      </p>
      <div className={styles.button}>
        <Button
          data-testid="add-policy-button"
          iconDescription="Create Policy"
          onClick={() => history.push(appLink.createPolicy({ teamId: activeTeamId }))}
          renderIcon={Add16}
          size="field"
        >
          Create Policy
        </Button>
      </div>
      {policies.length > 0 ? (
        <PoliciesTable policies={policies} />
      ) : (
        <NoDisplay text="No policies found. Go forth and create one or two why don't you?" style={{ width: "20rem" }} />
      )}
    </section>
  );
}

export default Policies;
