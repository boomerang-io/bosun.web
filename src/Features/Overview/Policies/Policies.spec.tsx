import React from "react";
import Policies from "./index";
import { policies as policiesData } from "ApiServer/fixtures";
const policies = [policiesData[1]];
describe("Policies --- Snapshot", () => {
  it("Capturing Snapshot of Policies", async () => {
    const { baseElement } = (global as any).rtlRouterRender(<Policies policies={policies} />);
    expect(baseElement).toMatchSnapshot();
  });
});
