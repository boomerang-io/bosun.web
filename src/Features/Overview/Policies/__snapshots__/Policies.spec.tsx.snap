// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policies --- Snapshot Capturing Snapshot of Policies 1`] = `
<body
  id="app"
>
  <div>
    <section
      class="container"
      data-testid="policies-container"
      id="policies-container"
    >
      <h2
        class="title"
      >
        Policies (1)
      </h2>
      <p
        class="message"
      >
        All of your teams policies can be found here. Go forth and define rules to enforce your standards!
      </p>
      <div
        class="button"
      >
        <button
          class="bx--btn bx--btn--field bx--btn--primary"
          data-testid="add-policy-button"
          tabindex="0"
          type="button"
        >
          Create Policy
          <svg
            aria-hidden="true"
            aria-label="Create Policy"
            class="bx--btn__icon"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            role="img"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
            />
          </svg>
        </button>
      </div>
      <div
        class="bx--data-table-container"
      >
        <div
          class="bx--data-table-content"
        >
          <table
            class="bx--data-table tableContainer bx--data-table--sort bx--data-table--no-border"
          >
            <thead>
              <tr
                class="tableHeadRow"
              >
                <th
                  aria-sort="none"
                  class="tableHeader name"
                  scope="col"
                >
                  <button
                    class="tableHeader name bx--table-sort"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Name
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeader definitions"
                  scope="col"
                >
                  <button
                    class="tableHeader definitions bx--table-sort"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Definitions
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeader rules"
                  scope="col"
                >
                  <button
                    class="tableHeader rules bx--table-sort"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Rules
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeader createdDate"
                  scope="col"
                >
                  <button
                    class="tableHeader createdDate bx--table-sort"
                    type="button"
                  >
                    <span
                      class="bx--table-sort__flex"
                    >
                      <div
                        class="bx--table-header-label"
                      >
                        Created Date
                      </div>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 30 17 30 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-label="Sort rows by this header in ascending order"
                        class="bx--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        role="img"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
              </tr>
            </thead>
            <tbody
              aria-live="polite"
              class="tableBody"
              data-testid="policies-tbody"
            >
              <tr
                class="tableRow"
                data-testid="policies-table-row"
              >
                <td
                  style="padding: 0px;"
                >
                  <div
                    class="tableCell"
                  >
                    <p
                      class="tableTextarea"
                    >
                      Timss
                    </p>
                  </div>
                </td>
                <td
                  style="padding: 0px;"
                >
                  <div
                    class="tableCell"
                  >
                    <p
                      class="tableTextarea"
                    >
                      2
                    </p>
                  </div>
                </td>
                <td
                  style="padding: 0px;"
                >
                  <div
                    class="tableCell"
                  >
                    <p
                      class="tableTextarea"
                    >
                      2
                    </p>
                  </div>
                </td>
                <td
                  style="padding: 0px;"
                >
                  <div
                    class="tableCell"
                  >
                    <time
                      class="tableTextarea"
                    >
                      January 01, 2020
                    </time>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </section>
  </div>
</body>
`;
